### ISSUE TITLE
[java] UseDiamondOperator false positive with var initializer

### ISSUE THREAD

**Issue by boris-petrov** [pmd/pmd#1624]
**Affects PMD Version:** 6.11.0

**Rule:** [UseDiamondOperator](https://pmd.github.io/pmd-6.11.0/pmd_rules_java_codestyle.html#usediamondoperator)

**Description:**

False positive when using "var": You cannot omit the `<String>` (if you want a typed array) but PMD complains.

**Code Sample demonstrating the issue:**

```java
var foo = new ArrayList<String>(5);
```

**Comment by adangel** [comment]
Thanks for reporting! That's clearly a false positive.

**Comment by Shubham-2k17** [comment]
I would like to work on this issue any hint on where to start will be helpful.

**Comment by jsotuyod** [comment]
@Shubham-2k17 sure!

As per [the docs](https://pmd.github.io/pmd-6.11.0/pmd_rules_java_codestyle.html#usediamondoperator) this is an XPath rule, defined by the following XPath:

```xpath
//VariableInitializer
//PrimaryExpression[not(PrimarySuffix)]
[not(ancestor::ArgumentList)]
/PrimaryPrefix/AllocationExpression[ClassOrInterfaceType[@AnonymousClass='false']/TypeArguments//ReferenceType[not(.//TypeArguments)]]
|
//StatementExpression[AssignmentOperator][PrimaryExpression/PrimaryPrefix[not(Expression)]]
//PrimaryExpression[not(PrimarySuffix)]
[not(ancestor::ArgumentList)]
/PrimaryPrefix/AllocationExpression[ClassOrInterfaceType[@AnonymousClass='false']/TypeArguments//ReferenceType[not(.//TypeArguments)]]
```

The "or" operator says this is basically 2 checks put together. One looks for assignment expressions in general, and the other look for variable initializers (that is, when a variable is assigned upon definition).

For Java 10+, `var` is allowed instead of the variable type of local vars, as long as *the variable initialization* allows to infer the variable type. This gives us two important pieces of information:

1. `var` can't be used without an initializer
2. only when using `var` we need full type info

So, given the code:

```java
var f = new ArrayList<String>(); // good! ArrayList<String> is the type we need to infer for f
f = ArrayList<>(); // good! f has already been defined, so here when the compiler finds <> it infers "String"
var s = new ArrayList<>(); // wrong! we don't know an array list of what is s!
var b; // wrong! we can't infer anything about the type of b!
```

So, we only need to look into the part of the XPath that checks initializers, and have it also check if it's using `var`; and if so, don't report anything.

The part of the expression checking for initializers is the first one:

```xpath
//VariableInitializer
//PrimaryExpression[not(PrimarySuffix)]
[not(ancestor::ArgumentList)]
/PrimaryPrefix/AllocationExpression[ClassOrInterfaceType[@AnonymousClass='false']/TypeArguments//ReferenceType[not(.//TypeArguments)]]
```

So we need to check where this initializer is being used.

Initializers can be in 2 places:
* array initializers (`ArrayInitializer`)
* a local / field variable declaration (`VariableDeclarator`, which is composed of `VariableDeclaratorId + VariableInitializer`)

If it is a `VariableDeclarator`, then we should check if the accompaning `VariableDeclaratorId` has the attribute `isTypeInferred` set to true.

So, basically, you should add a check right after `//VariableInitializer`, to check that the ancestor is either not a `VariableDeclarator`, or the `isTypeInferred` (`@typeInferred` under XPath) of the sibling `VariableDeclaratorId` is false.

Using the designer is key to getting this XPath fixed. It will allow you to edit the XPath and test it on the fly.

Once you have the expression updated, change it in the [code style ruleset](https://github.com/pmd/pmd/blob/cfe07822be49311797a539a9b3cffb4de4fb512e/pmd-java/src/main/resources/category/java/codestyle.xml#L1954-L1962), and add the proper test cases in the XML as with any other rule you have done so far.

**Comment by Shubham-2k17** [comment]
Hi @jsotuyod ,
I have some doubts regarding one of the test case. After modifying the XPATH expression which is working fine for var and other test cases except this one:
```
import java.util.ArrayList;
import java.util.List;
public class Foo {
    List<String> field;
    public void foo() {
       List<String> strings = new ArrayList<String>();
       List<String> strings2 = new ArrayList<>();
       this.field = new ArrayList<String>();
    }
}
```
PMD is flagging line number **6** and **8** but test case is expecting line number **8** and **11**.
6 and 8 are the correct line numbers. Is the test case wrong or i am missing something?

**Comment by Shubham-2k17** [comment]
I changed an existing test case in this PR. Do check if it is correct.
